src/Psl/Type/Internal/BackedEnumValueType.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use BackedEnum;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\Type;
+
+use function is_int;
+use function is_string;
+use function Psl\Type\int;
+use function Psl\Type\string;
+
+/**
+ * @template T of BackedEnum
+ *
+ * @extends Type<value-of<T>>
+ *
+ * @internal
+ */
+final readonly class BackedEnumValueType extends Type
+{
+    private bool $isStringBacked;
+
+    /**
+     * @psalm-mutation-free
+     *
+     * @param class-string<T> $enum
+     */
+    public function __construct(
+        private string $enum
+    ) {
+        $this->isStringBacked = is_string($this->enum::cases()[0]->value);
```

Comment:
This code will produce warnings if the enum has no values.

e.g.

```php
enum Bar: string {}

\Psl\Type\backed_enum_value(Bar::class);
```

Not sure if we want to validate against that, or if its not worth it.

Reply:
we should use reflections here.

Reply:
I didn't think it was worth the penalty of using reflection, but I've changed it to reflect the enum now 👍

Reply:
Doing some quick checks on my machine, reflection is a factor 3 slower. Maybe we could fall back to it only if there are no cases?

Reply:
OK, I've extracted detection to a private method to make the purpose of the added code obvious and added additional tests to cover the detection logic.

Reply:
I'm also unsure how to force a reflection failure here to test the uncovered lines. Any ideas?